File: src/services/github.ts
import { Octokit } from '@octokit/rest';
import { ValidationError } from '../lib/errors.js';
import { outputWarning } from './output.js';
import type { CommitSummary, PRSummary, ReleaseSummary } from '../models/observation.js';

/**
 * Create an authenticated Octokit instance.
 */
export function createOctokit(token: string): Octokit {
  return new Octokit({ auth: token });
}

/**
 * Extract owner and repo name from a GitHub URL.
 *
 * Supports:
 *   - https://github.com/owner/repo
 *   - https://github.com/owner/repo.git
 *   - git@github.com:owner/repo.git
 */
export function parseRepoUrl(url: string): { owner: string; repo: string } {
  // HTTPS format
  const httpsMatch = url.match(/github\.com\/([^/]+)\/([^/.]+)(?:\.git)?/);
  if (httpsMatch) {
    return { owner: httpsMatch[1], repo: httpsMatch[2] };
  }

  // SSH format
  const sshMatch = url.match(/github\.com:([^/]+)\/([^/.]+)(?:\.git)?/);
  if (sshMatch) {
    return { owner: sshMatch[1], repo: sshMatch[2] };
  }

  throw new ValidationError(`Cannot parse GitHub owner/repo from URL: ${url}`);
}

/**
 * Check rate-limit headers and warn if running low.
 */
function checkRateLimit(headers: Record<string, string | number | undefined>): void {
  const remaining = headers['x-ratelimit-remaining'];
  if (remaining !== undefined && Number(remaining) < 100) {
    outputWarning(`GitHub API rate limit low: ${remaining} requests remaining.`);
  }
}

/**
 * Fetch commits from a repository, optionally since a given date.
 */
export async function fetchCommits(
  octokit: Octokit,
  owner: string,
  repo: string,
  since?: Date,
): Promise<CommitSummary[]> {
  const params: Record<string, unknown> = {
    owner,
    repo,
    per_page: 100,
  };
  if (since) {
    params.since = since.toISOString();
  }

  const response = await octokit.repos.listCommits(params as Parameters<typeof octokit.repos.listCommits>[0]);
  checkRateLimit(response.headers as Record<string, string | number | undefined>);

  return response.data.map((c) => ({
    sha: c.sha,
    message: c.commit.message.split('\n')[0],
    author: c.commit.author?.name ?? c.author?.login ?? 'unknown',
    date: c.commit.author?.date ?? '',
  }));
}

/**
 * Fetch pull requests from a repository, optionally filtered by date.
 */
export async function fetchPullRequests(
  octokit: Octokit,
  owner: string,
  repo: string,
  since?: Date,
): Promise<PRSummary[]> {
  const response = await octokit.pulls.list({
    owner,
    repo,
    state: 'all',
    sort: 'updated',
    direction: 'desc',
    per_page: 100,
  });
  checkRateLimit(response.headers as Record<string, string | number | undefined>);

  let prs = response.data;
  if (since) {
    const sinceTime = since.getTime();
    prs = prs.filter((pr) => new Date(pr.updated_at).getTime() >= sinceTime);
  }

  return prs.map((pr) => ({
    number: pr.number,
    title: pr.title,
    state: pr.state,
    author: pr.user?.login ?? 'unknown',
    url: pr.html_url,
  }));
}

/**
 * Fetch releases from a repository, optionally filtered by date.
 */
export async function fetchReleases(
  octokit: Octokit,
  owner: string,
  repo: string,
  since?: Date,
): Promise<ReleaseSummary[]> {
  const response = await octokit.repos.listReleases({
    owner,
    repo,
    per_page: 100,
  });
  checkRateLimit(response.headers as Record<string, string | number | undefined>);

  let releases = response.data;
  if (since) {
    const sinceTime = since.getTime();
    releases = releases.filter(
      (r) => r.published_at && new Date(r.published_at).getTime() >= sinceTime,
    );
  }

  return releases.map((r) => ({
    tag: r.tag_name,
    name: r.name ?? r.tag_name,
    date: r.published_at ?? r.created_at,
    url: r.html_url,
  }));
}

/**
 * Fetch diff information for a specific pull request.
 */
export async function fetchPRDiff(
  octokit: Octokit,
  owner: string,
  repo: string,
  prNumber: number,
): Promise<{ files: string[]; additions: number; deletions: number; patch: string }> {
  const [prResponse, filesResponse] = await Promise.all([
    octokit.pulls.get({ owner, repo, pull_number: prNumber }),
    octokit.pulls.listFiles({ owner, repo, pull_number: prNumber, per_page: 100 }),
  ]);

  checkRateLimit(prResponse.headers as Record<string, string | number | undefined>);

  const files = filesResponse.data.map((f) => f.filename);
  const additions = filesResponse.data.reduce((sum, f) => sum + f.additions, 0);
  const deletions = filesResponse.data.reduce((sum, f) => sum + f.deletions, 0);
  const patch = filesResponse.data.map((f) => f.patch ?? '').join('\n');

  return { files, additions, deletions, patch };
}

/**
 * Fetch diff information for a specific commit.
 */
export async function fetchCommitDiff(
  octokit: Octokit,
  owner: string,
  repo: string,
  sha: string,
): Promise<{ files: string[]; additions: number; deletions: number; message: string }> {
  const response = await octokit.repos.getCommit({ owner, repo, ref: sha });
  checkRateLimit(response.headers as Record<string, string | number | undefined>);

  const commitFiles = response.data.files ?? [];
  const files = commitFiles.map((f) => f.filename);
  const additions = commitFiles.reduce((sum, f) => sum + f.additions, 0);
  const deletions = commitFiles.reduce((sum, f) => sum + f.deletions, 0);
  const message = response.data.commit.message;

  return { files, additions, deletions, message };
}

File: code_adapt/services/github.py
"""GitHub REST API client via httpx."""

from __future__ import annotations

import re
from datetime import datetime

import httpx

from ..errors import ValidationError
from ..models import CommitSummary, PRSummary, ReleaseSummary

API_BASE = "https://api.github.com"


def parse_repo_url(url: str) -> tuple[str, str]:
    """Extract (owner, repo) from a GitHub URL (HTTPS or SSH)."""
    m = re.search(r"github\.com[/:]([^/]+)/([^/.]+?)(?:\.git)?$", url)
    if not m:
        raise ValidationError(f"Cannot parse GitHub owner/repo from URL: {url}")
    return m.group(1), m.group(2)


def _client(token: str) -> httpx.Client:
    return httpx.Client(
        base_url=API_BASE,
        headers={
            "Authorization": f"Bearer {token}",
            "Accept": "application/vnd.github+json",
            "X-GitHub-Api-Version": "2022-11-28",
        },
        timeout=30.0,
    )


def _check_rate_limit(resp: httpx.Response) -> None:
    remaining = resp.headers.get("x-ratelimit-remaining")
    if remaining is not None and int(remaining) < 100:
        from rich import print as rprint

        rprint(f"[yellow]GitHub API rate limit low: {remaining} requests remaining.[/yellow]")


def fetch_commits(
    token: str, owner: str, repo: str, since: datetime | None = None
) -> list[CommitSummary]:
    with _client(token) as client:
        params: dict = {"per_page": 100}
        if since:
            params["since"] = since.isoformat()
        resp = client.get(f"/repos/{owner}/{repo}/commits", params=params)
        resp.raise_for_status()
        _check_rate_limit(resp)
        return [
            CommitSummary(
                sha=c["sha"],
                message=c["commit"]["message"].split("\n")[0],
                author=c["commit"]["author"]["name"]
                if c["commit"]["author"]
                else c.get("author", {}).get("login", "unknown"),
                date=c["commit"]["author"]["date"] if c["commit"]["author"] else "",
            )
            for c in resp.json()
        ]


def fetch_pull_requests(
    token: str, owner: str, repo: str, since: datetime | None = None
) -> list[PRSummary]:
    with _client(token) as client:
        resp = client.get(
            f"/repos/{owner}/{repo}/pulls",
            params={"state": "all", "sort": "updated", "direction": "desc", "per_page": 100},
        )
        resp.raise_for_status()
        _check_rate_limit(resp)
        prs = resp.json()
        if since:
            since_ts = since.timestamp()
            prs = [
                pr
                for pr in prs
                if datetime.fromisoformat(pr["updated_at"].replace("Z", "+00:00")).timestamp()
                >= since_ts
            ]
        return [
            PRSummary(
                number=pr["number"],
                title=pr["title"],
                state=pr["state"],
                author=pr.get("user", {}).get("login", "unknown"),
                url=pr["html_url"],
            )
            for pr in prs
        ]


def fetch_releases(
    token: str, owner: str, repo: str, since: datetime | None = None
) -> list[ReleaseSummary]:
    with _client(token) as client:
        resp = client.get(f"/repos/{owner}/{repo}/releases", params={"per_page": 100})
        resp.raise_for_status()
        _check_rate_limit(resp)
        releases = resp.json()
        if since:
            since_ts = since.timestamp()
            releases = [
                r
                for r in releases
                if r.get("published_at")
                and datetime.fromisoformat(
                    r["published_at"].replace("Z", "+00:00")
                ).timestamp()
                >= since_ts
            ]
        return [
            ReleaseSummary(
                tag=r["tag_name"],
                name=r.get("name") or r["tag_name"],
                date=r.get("published_at") or r["created_at"],
                url=r["html_url"],
            )
            for r in releases
        ]


def fetch_pr_diff(
    token: str, owner: str, repo: str, pr_number: int
) -> dict:
    """Return {files, additions, deletions, patch}."""
    with _client(token) as client:
        files_resp = client.get(
            f"/repos/{owner}/{repo}/pulls/{pr_number}/files", params={"per_page": 100}
        )
        files_resp.raise_for_status()
        _check_rate_limit(files_resp)
        file_data = files_resp.json()

        pr_resp = client.get(f"/repos/{owner}/{repo}/pulls/{pr_number}")
        pr_resp.raise_for_status()
        title = pr_resp.json()["title"]

        files = [f["filename"] for f in file_data]
        additions = sum(f["additions"] for f in file_data)
        deletions = sum(f["deletions"] for f in file_data)
        return {
            "files": files,
            "additions": additions,
            "deletions": deletions,
            "message": title,
        }


def fetch_commit_diff(
    token: str, owner: str, repo: str, sha: str
) -> dict:
    """Return {files, additions, deletions, message}."""
    with _client(token) as client:
        resp = client.get(f"/repos/{owner}/{repo}/commits/{sha}")
        resp.raise_for_status()
        _check_rate_limit(resp)
        data = resp.json()
        commit_files = data.get("files", [])
        return {
            "files": [f["filename"] for f in commit_files],
            "additions": sum(f["additions"] for f in commit_files),
            "deletions": sum(f["deletions"] for f in commit_files),
            "message": data["commit"]["message"],
        }


def fetch_release_info(token: str, owner: str, repo: str, tag: str) -> dict:
    """Return {name, message}."""
    with _client(token) as client:
        resp = client.get(f"/repos/{owner}/{repo}/releases/tags/{tag}")
        resp.raise_for_status()
        data = resp.json()
        return {"name": data.get("name") or tag, "message": data.get("name") or tag}
